# Social Network Data Visualization

This projects builds upon the previous one we completed: "Classifying Nobel Prize Physicists by Physics Field."

In that project, we assigned each Nobel Prize Physicist to one of 6 physics subfields. Now, we graphically represent all these physicists (along with various physics domains) as a social network.

## Social network - The basics
A social network consists of nodes and links between those nodes. We produce a small social network as a json file, where the nodes are phsicists, and the links are "connections between them". The file produced is named "small_network.json". To visualize this network, you must open the "small_network.html" file (provided by Damien Benveniste) in your preferred browser. Both the JSON and HTML files are available in the repository.

NOTE: In order to work, the json file MUST be in a folder titled "data", which is itself within the folder containing the html file.

In [1]:
small_network = {
    "nodes": [
        {"id": "Albert Einstein"},
        {"id": "Paul Dirac"},
        {"id": "Niels Bohr"}
    ],
    "links": [
        {"source": "Albert Einstein", "target": "Paul Dirac"},
        {"source": "Albert Einstein", "target": "Niels Bohr"},
        {"source": "Paul Dirac", "target": "Niels Bohr"}
    ]
}

## We dump this network into a .json file
import json
with open(".\data\small_network.json","w") as f:
    json.dump(small_network, f, indent=4)

The networks just created is simple. The rest of this project builds a more complex social network, connecting Nobel physicists and various physics topics.

## Getting physicists and physics domains data

We are going to gather the words from the Wikipedia pages for physicists and physics domains. This will be the information we use to determine which physicists/domains are connect to which, and how strong each connection is.

We first get the links for each Nobel physicist's and physics domains's Wikipedia page. We identify the Nobel physicists by a table present at "https://en.wikipedia.org/wiki/List_of_Nobel_laureates_in_Physics" (as of 12 Jul 2018) which lists them all. We get the physics topics from the Reasearch table at "https://en.wikipedia.org/wiki/Physics", which lists all these physics domains.

- ### Gathering physicists links
We gather the links for the physicists. This is largely done using the code from the previous project, "Classifying Nobel Prize Physicists by Physics Field." However, we drop the Year, Country, Rationale that we gather. We then use Beautiful Soup to parse for the links on Wikipedia. At the end of this we have a df with the physicist ("Laureate") as the index, and containing the links for each physicist's Wikipedia page.

In [3]:
## We get the nobel data set
import numpy as np
import pandas as pd
from httplib2 import Http
from bs4 import BeautifulSoup, SoupStrainer

class Parser:
    
    def __init__(self, url):  
        http = Http()
        status, response = http.request(url)
        tables = BeautifulSoup(response, "lxml", 
                              parse_only=SoupStrainer("table", {"class":"wikitable plainrowheaders sortable"}))
        self.table = tables.contents[1]
    
    def parse_table(self):      
        rows = self.table.find_all("tr")
        header = self.parse_header(rows[0])
        table_array = [self.parse_row(row) for row in rows[1:]]
        table_df = pd.DataFrame(table_array, columns=header).apply(self.clean_table, 1)
        table_df = table_df.replace({"Year":{'':np.nan}})
        table_df.drop(labels = ['Ref','Image'], axis =1, inplace=True)
        return table_df
        
    def parse_row(self, row):     
        columns = row.find_all(["td","th"])
        return [BeautifulSoup.get_text(col).strip() for col in columns if BeautifulSoup.get_text(col) != '']
    
    def parse_header(self, row):     
        columns = row.find_all("th")
        to_return = [BeautifulSoup.get_text(col).strip() for col in columns if BeautifulSoup.get_text(col) != ""]
        return to_return

    def clean_table(self, row):
        if not row.iloc[0].isdigit():
            return row.shift(periods = 1)
        else:
            return row
        
url = "https://en.wikipedia.org/wiki/List_of_Nobel_laureates_in_Physics"        
parser = Parser(url)   
nobel_df = parser.parse_table()

nobel_df.columns = ["Year", "Laureate", "Country", "Rationale"]
nobel_df.dropna(subset=["Country"], inplace=True)
nobel_df.fillna(method="ffill", inplace=True)
nobel_df.drop(["Year", "Country", "Rationale"], 1, inplace=True)

http = Http()
status, response = http.request(url)

table = BeautifulSoup(response, "lxml", parse_only=SoupStrainer('table'))
link_df = pd.DataFrame([[x.string, x["href"]] for x in table.contents[1].find_all("a")],
                       columns=["Laureate", "link"]).drop_duplicates()

nobel_df = nobel_df.merge(link_df, on="Laureate", how="left")
nobel_df.set_index("Laureate", inplace=True)
nobel_df.drop_duplicates(inplace=True)
nobel_df

link
Laureate                                                                      
Wilhelm Conrad Röntgen                              /wiki/Wilhelm_R%C3%B6ntgen
Hendrik Lorentz                                          /wiki/Hendrik_Lorentz
Pieter Zeeman                                              /wiki/Pieter_Zeeman
Antoine Henri Becquerel                                  /wiki/Henri_Becquerel
Pierre Curie                                                /wiki/Pierre_Curie
Maria Skłodowska-Curie                       /wiki/Maria_Sk%C5%82odowska-Curie
Lord Rayleigh                    /wiki/John_William_Strutt,_3rd_Baron_Rayleigh
Philipp Eduard Anton von Lenard                           /wiki/Philipp_Lenard
Joseph John Thomson                                        /wiki/J._J._Thomson
Albert Abraham Michelson                             /wiki/Albert_A._Michelson
Gabriel Lippmann                                        /wiki/Gabriel_Lippmann
Guglielmo Marconi                                      /wiki/Guglielmo_Marconi
Karl Ferdinand Braun                                /wiki/Karl_Ferdinand_Braun
Johannes Diderik van der Waals            /wiki/Johannes_Diderik_van_der_Waals
Wilhelm Wien                                                /wiki/Wilhelm_Wien
Nils Gustaf Dalén                                      /wiki/Gustaf_Dal%C3%A9n
Heike Kamerlingh-Onnes                            /wiki/Heike_Kamerlingh_Onnes
Max von Laue                                                /wiki/Max_von_Laue
William Henry Bragg                                  /wiki/William_Henry_Bragg
William Lawrence Bragg                                    /wiki/Lawrence_Bragg
Charles Glover Barkla                              /wiki/Charles_Glover_Barkla
Max Planck                                                    /wiki/Max_Planck
Johannes Stark                                            /wiki/Johannes_Stark
Charles Édouard Guillaume                 /wiki/Charles_%C3%89douard_Guillaume
Albert Einstein                                          /wiki/Albert_Einstein
Niels Bohr                                                    /wiki/Niels_Bohr
Robert Andrews Millikan                          /wiki/Robert_Andrews_Millikan
Manne Siegbahn                                            /wiki/Manne_Siegbahn
James Franck                                                /wiki/James_Franck
Gustav Hertz                                         /wiki/Gustav_Ludwig_Hertz
...                                                                        ...
John C. Mather                                            /wiki/John_C._Mather
George F. Smoot                                             /wiki/George_Smoot
Albert Fert                                                  /wiki/Albert_Fert
Peter Grünberg                                       /wiki/Peter_Gr%C3%BCnberg
Makoto Kobayashi                            /wiki/Makoto_Kobayashi_(physicist)
Toshihide Maskawa                                      /wiki/Toshihide_Maskawa
Yoichiro Nambu                                            /wiki/Yoichiro_Nambu
Charles K. Kao                                            /wiki/Charles_K._Kao
Willard S. Boyle                                           /wiki/Willard_Boyle
George E. Smith                                          /wiki/George_E._Smith
Andre Geim                                                    /wiki/Andre_Geim
Konstantin Novoselov                                /wiki/Konstantin_Novoselov
Saul Perlmutter                                          /wiki/Saul_Perlmutter
Brian P. Schmidt                                           /wiki/Brian_Schmidt
Adam G. Riess                                                 /wiki/Adam_Riess
Serge Haroche                                              /wiki/Serge_Haroche
David J. Wineland                                      /wiki/David_J._Wineland
François Englert                                   /wiki/Fran%C3%A7ois_Englert
Peter Higgs                                  

- ### Gather physics domain links

Using similar code, and taking advantage of Beautiful Soup, we gather the links for the physics domains, and place them in a df named "physics_df".

In [20]:
## We get the physics links
url = "https://en.wikipedia.org/wiki/Physics"

http = Http()
status, response = http.request(url)

table = BeautifulSoup(response, "lxml", parse_only=SoupStrainer('table'))
physics_df = pd.DataFrame([[x.string.lower(), x["href"].lower()] for x in table.contents[3].find_all("a")],
                       columns=["Physics_domain", "link"]).drop_duplicates()

physics_df = physics_df.groupby("Physics_domain").first()

In [21]:
physics_df

link
Physics_domain                                                                      
accelerator physics                                        /wiki/accelerator_physics
acoustics                                                            /wiki/acoustics
agrophysics                                                        /wiki/agrophysics
antimatter                                                          /wiki/antimatter
applied physics                                                /wiki/applied_physics
astrometry                                                          /wiki/astrometry
astronomy                                                            /wiki/astronomy
astrophysics                                                      /wiki/astrophysics
atom                                                                      /wiki/atom
atomic and molecular astrophysics            /wiki/atomic_and_molecular_astrophysics
atomic physics                                                  /wiki/atomic_physics
atomic, molecular, and optical physics  /wiki/atomic,_molecular,_and_optical_physics
bcs theory                                                          /wiki/bcs_theory
big bang                                                              /wiki/big_bang
biophysics                                                          /wiki/biophysics
black hole                                                          /wiki/black_hole
bloch wave                                                          /wiki/bloch_wave
bose-einstein condensate                              /wiki/bose-einstein_condensate
brane                                                      /wiki/membrane_(m-theory)
casimir effect                                                  /wiki/casimir_effect
chemical physics                                              /wiki/chemical_physics
communication physics                                    /wiki/communication_physics
condensed matter physics                              /wiki/condensed_matter_physics
cosmic background radiation                        /wiki/cosmic_background_radiation
cosmic inflation                                              /wiki/cosmic_inflation
cosmic string                                                    /wiki/cosmic_string
cosmology                                                   /wiki/physical_cosmology
cosmos                                                                  /wiki/cosmos
dark energy                                                        /wiki/dark_energy
dark matter                                                        /wiki/dark_matter
...                                                                              ...
quantum optics                                                  /wiki/quantum_optics
seesaw mechanism                                              /wiki/seesaw_mechanism
self-organization                                            /wiki/self-organization
semiconductor                                                    /wiki/semiconductor
solar physics                                                    /wiki/solar_physics
solar system                                                      /wiki/solar_system
solid                                                                    /wiki/solid
solid state physics                                        /wiki/solid_state_physics
solid-state devices                                        /wiki/solid-state_physics
space physics                                                    /wiki/space_physics
spectral line                                                    /wiki/spectral_line
spin                                                            /wiki/spin_(physics)
spontaneous symmetry breaking                    /wiki/spontaneous_symmetry_breaking
standard model                                                  /wiki/standard_model
star                                                                      /wiki/star


## Gather important words from Wikipedia pages

Now that we have the links, we need the words from each Wikipedia page corresponding to each link. We define a function to get a text, and multiple functions to keep only those words we consider "important". Largely, this code is borrowed from our previous project, "Classifying Nobel Prize Physicists by Physics Field." One difference here is that we aggregrate all these functions into a single function called "clean_everything" that will gather AND clean the words.

- ### Gather all words from Wikipedia bios

In [22]:
from string import punctuation

## We get the bios
def get_text(link, root_website = "https://en.wikipedia.org"):    
    http = Http()
    status, response = http.request(root_website + link)
    body = BeautifulSoup(response, "lxml", parse_only=SoupStrainer("div", {"id":"mw-content-text"}))
    return BeautifulSoup.get_text(body.contents[1])

- ### Copy clean_string, remove, remove_one from previous project

In [30]:
## Copy clean_string function from Nobel Classification project
def clean_string(string):
    for p in punctuation + "1234567890":
        string = string.replace(p,'').lower()
    return string

## Copy remove function from the Nobel Classification project
def remove(list_to_clean, element_to_remove=[None, ""]):
    list_cleaned = [x for x in list_to_clean if x not in element_to_remove]
    return list_cleaned

## Copy remove_one function from Nobel Classification project
def remove_one(list_to_clean):
    list_cleaned = [x for x in list_to_clean if len(x) > 1]
    return list_cleaned

## Import stop words to remove
from nltk.corpus import stopwords
words_to_remove = set(stopwords.words('english'))

- ### Aggregate all the above functions into one to return a list of words from each link

In [31]:
def clean_everything(df):
    fin_list = df['link'].apply(get_text).apply(clean_string).str.split().apply(remove).apply(remove, element_to_remove = words_to_remove).apply(remove_one)
    return fin_list

physics_df["physics_list"] = clean_everything(physics_df)
nobel_df["physics_list"] = clean_everything(nobel_df)
nobel_df

link  \
Laureate                                                                         
Wilhelm Conrad Röntgen                              /wiki/Wilhelm_R%C3%B6ntgen   
Hendrik Lorentz                                          /wiki/Hendrik_Lorentz   
Pieter Zeeman                                              /wiki/Pieter_Zeeman   
Antoine Henri Becquerel                                  /wiki/Henri_Becquerel   
Pierre Curie                                                /wiki/Pierre_Curie   
Maria Skłodowska-Curie                       /wiki/Maria_Sk%C5%82odowska-Curie   
Lord Rayleigh                    /wiki/John_William_Strutt,_3rd_Baron_Rayleigh   
Philipp Eduard Anton von Lenard                           /wiki/Philipp_Lenard   
Joseph John Thomson                                        /wiki/J._J._Thomson   
Albert Abraham Michelson                             /wiki/Albert_A._Michelson   
Gabriel Lippmann                                        /wiki/Gabriel_Lippmann   
Guglielmo Marconi                                      /wiki/Guglielmo_Marconi   
Karl Ferdinand Braun                                /wiki/Karl_Ferdinand_Braun   
Johannes Diderik van der Waals            /wiki/Johannes_Diderik_van_der_Waals   
Wilhelm Wien                                                /wiki/Wilhelm_Wien   
Nils Gustaf Dalén                                      /wiki/Gustaf_Dal%C3%A9n   
Heike Kamerlingh-Onnes                            /wiki/Heike_Kamerlingh_Onnes   
Max von Laue                                                /wiki/Max_von_Laue   
William Henry Bragg                                  /wiki/William_Henry_Bragg   
William Lawrence Bragg                                    /wiki/Lawrence_Bragg   
Charles Glover Barkla                              /wiki/Charles_Glover_Barkla   
Max Planck                                                    /wiki/Max_Planck   
Johannes Stark                                            /wiki/Johannes_Stark   
Charles Édouard Guillaume                 /wiki/Charles_%C3%89douard_Guillaume   
Albert Einstein                                          /wiki/Albert_Einstein   
Niels Bohr                                                    /wiki/Niels_Bohr   
Robert Andrews Millikan                          /wiki/Robert_Andrews_Millikan   
Manne Siegbahn                                            /wiki/Manne_Siegbahn   
James Franck                                                /wiki/James_Franck   
Gustav Hertz                                         /wiki/Gustav_Ludwig_Hertz   
...                                                                        ...   
John C. Mather                                            /wiki/John_C._Mather   
George F. Smoot                                             /wiki/George_Smoot   
Albert Fert                                                  /wiki/Albert_Fert   
Peter Grünberg                                       /wiki/Peter_Gr%C3%BCnberg   
Makoto Kobayashi                            /wiki/Makoto_Kobayashi_(physicist)   
Toshihide Maskawa                                      /wiki/Toshihide_Maskawa   
Yoichiro Nambu                                            /wiki/Yoichiro_Nambu   
Charles K. Kao                                            /wiki/Charles_K._Kao   
Willard S. Boyle                                           /wiki/Willard_Boyle   
George E. Smith                                          /wiki/George_E._Smith   
Andre Geim                                                    /wiki/Andre_Geim   
Konstantin Novoselov                                /wiki/Konstantin_Novoselov   
Saul Perlmutter                                          /wiki/Saul_Perlmutter   
Brian P. Schmidt                                           /wiki/Brian_Schmidt   
Adam G. Riess                                                 /wiki/Adam_Riess   
Serge Haroche                                              /wiki/Serge_Haroche   
David J. Wineland                                      /

## How many words?

We now have all the words from the Wikipedia pages for the physicists and physics domains. Let's find out how many unique words are in the nobel_df["physics_list"], how many are in physics_df["physics_list"], how many are in the interesction of the two, and how many of the words in each entry (bio) of each respective list are in the intersection.

- ### Find all the words in nobel_df["physics_list"]

In [32]:
all_nobel_words =  set( nobel_df["physics_list"].sum() )
all_nobel_words

{'camilla',
 'buckley',
 'température',
 'angelesdoctoral',
 'thickness',
 'breslow',
 'northen',
 'armistice',
 'distorted',
 'maldon',
 'sampson',
 'howard',
 'maquoketa',
 'schik',
 'awarding',
 'judd',
 'mussolini',
 'interpretational',
 'driggs',
 'longest',
 'europhysics',
 'uhuru',
 'kaka',
 'honesttogoodness',
 'nicéphore',
 'colman',
 'delhi',
 'cars',
 'conducting',
 'hbs',
 'человека',
 'deserlawrence',
 'darin',
 'bscknown',
 'marshakgordon',
 'bombed',
 'cancellation',
 'anna',
 'roger',
 'ovation',
 'kattowitz',
 'whhmxg',
 'romanian',
 'ccting',
 'donor',
 'expertise',
 'distortion',
 'australiancitizenship',
 '秀樹born',
 'distort',
 'danby',
 'lazarevich',
 'orb',
 'shiver',
 'arxivhepex',
 'indexes',
 'watts',
 'www',
 'chap',
 'jrfebruary',
 'kölliker',
 'rijkedoctoral',
 'peres',
 'poets',
 'meti',
 'nandalal',
 'byrobert',
 'rich',
 'obsolescent',
 'alton',
 'arthurs',
 'unassailability',
 'pseudoreligious',
 'chidambaram',
 'bacterial',
 'isotopesmass',
 'boyle',
 '

- ### Find all the words in physics_df["physics_list"]

In [33]:
all_physics_words =  set( physics_df["physics_list"].sum() )
all_physics_words

{'paints',
 'deimosmoon',
 'galyam',
 'debyes',
 'thickness',
 'logicians',
 'distorted',
 'sculptures',
 'howard',
 'green–kubo',
 'chevallier',
 'cαγo',
 'cgh',
 'longest',
 'patil',
 'europhysics',
 'florian',
 'motioncitation',
 'conceptedit',
 'subbrown',
 'cars',
 'conducting',
 'bina',
 'sushruta',
 'cancellation',
 'meanfield',
 'eccentricity',
 'maire',
 'anna',
 'roger',
 'ldx',
 'lamarckism',
 'romanian',
 'offshore',
 'neoncopper',
 'brandon',
 'setend',
 'slicing',
 'expertise',
 'donor',
 'distortion',
 'distort',
 'danby',
 'rotatingnonspherical',
 'sipylum',
 'orb',
 'alhazens',
 'superdense',
 'qm',
 'arxivhepex',
 'watts',
 'grummans',
 'indexes',
 'riser',
 'www',
 'chap',
 'integrator',
 'decker',
 'peres',
 'trick',
 'tunnelingcurrent',
 'straightforwardly',
 'rich',
 'snr',
 'cladding',
 'leftcidagger',
 'biomaterials',
 'bk†',
 'deviates',
 'bacterial',
 'bangedit',
 'orientation',
 'even–even',
 'boyle',
 'acs',
 'lose',
 'samples',
 'floatovoltaics',
 'timetrav

- ### Find the intersection of all_nobel_words and all_physics_words

In [34]:
physics_corpus =  all_nobel_words.intersection(all_physics_words)

physics_corpus

{'paved',
 'proposal',
 'released',
 'wwwencyclopediacom',
 'pbar',
 'exact',
 'mystic',
 'thickness',
 'element',
 'aas',
 'poisonous',
 'ralph',
 'oleary',
 'distorted',
 'aqueous',
 'howard',
 'society',
 'reject',
 'le',
 'ramsey',
 'students',
 'liberated',
 'jaguar',
 'mother',
 'wrinkles',
 'kids',
 'longest',
 'europhysics',
 'electrochemical',
 'abelian',
 'religious',
 'natura',
 'detroit',
 'bibcodephrvk',
 'unanticipated',
 'enumerating',
 'dem',
 'nearer',
 'thermostatic',
 'cars',
 'conducting',
 'keeps',
 'genes',
 'intrigued',
 'allied',
 'stripped',
 'dj',
 'automaton',
 'eta',
 'harwood',
 'manifest',
 'cancellation',
 'contain',
 'stabilized',
 'trend',
 'installations',
 'lists',
 'anna',
 'roger',
 'superconductive',
 'glasses',
 'romanian',
 'pearce',
 'highquality',
 'schopenhauer',
 'donor',
 'expertise',
 'distortion',
 'chu',
 'congress',
 'manmade',
 'drawn',
 'routes',
 'distort',
 'danby',
 'alkali',
 'orb',
 'preferring',
 'ring',
 'microscope',
 'bibcoden

In [35]:
print(len(physics_corpus), len(all_nobel_words), len(all_physics_words))

12935 33378 34722


So the unique words from the physicist pages total 33,378.
The unique words from the physics domain pages totals 34,722.
The unique words they share in common is 12,935. A little more than a third of the total words in either.

For every entry in the "physics_list" for both nobel_df and physics_df, we are going to keep only those words in the intersection (physics_corpus, as defined above). We will store these in a new column in our dataframes, "physics_list_clean".

- ### Keep only words in physics corpus.

In [36]:
def keep_only(list_to_clean, corpus=physics_corpus):
    return [word for word in list_to_clean if word in corpus]
    
nobel_df["physics_list_clean"] = nobel_df["physics_list"].apply(keep_only)
physics_df["physics_list_clean"] = physics_df["physics_list"].apply(keep_only)
nobel_df

link  \
Laureate                                                                         
Wilhelm Conrad Röntgen                              /wiki/Wilhelm_R%C3%B6ntgen   
Hendrik Lorentz                                          /wiki/Hendrik_Lorentz   
Pieter Zeeman                                              /wiki/Pieter_Zeeman   
Antoine Henri Becquerel                                  /wiki/Henri_Becquerel   
Pierre Curie                                                /wiki/Pierre_Curie   
Maria Skłodowska-Curie                       /wiki/Maria_Sk%C5%82odowska-Curie   
Lord Rayleigh                    /wiki/John_William_Strutt,_3rd_Baron_Rayleigh   
Philipp Eduard Anton von Lenard                           /wiki/Philipp_Lenard   
Joseph John Thomson                                        /wiki/J._J._Thomson   
Albert Abraham Michelson                             /wiki/Albert_A._Michelson   
Gabriel Lippmann                                        /wiki/Gabriel_Lippmann   
Guglielmo Marconi                                      /wiki/Guglielmo_Marconi   
Karl Ferdinand Braun                                /wiki/Karl_Ferdinand_Braun   
Johannes Diderik van der Waals            /wiki/Johannes_Diderik_van_der_Waals   
Wilhelm Wien                                                /wiki/Wilhelm_Wien   
Nils Gustaf Dalén                                      /wiki/Gustaf_Dal%C3%A9n   
Heike Kamerlingh-Onnes                            /wiki/Heike_Kamerlingh_Onnes   
Max von Laue                                                /wiki/Max_von_Laue   
William Henry Bragg                                  /wiki/William_Henry_Bragg   
William Lawrence Bragg                                    /wiki/Lawrence_Bragg   
Charles Glover Barkla                              /wiki/Charles_Glover_Barkla   
Max Planck                                                    /wiki/Max_Planck   
Johannes Stark                                            /wiki/Johannes_Stark   
Charles Édouard Guillaume                 /wiki/Charles_%C3%89douard_Guillaume   
Albert Einstein                                          /wiki/Albert_Einstein   
Niels Bohr                                                    /wiki/Niels_Bohr   
Robert Andrews Millikan                          /wiki/Robert_Andrews_Millikan   
Manne Siegbahn                                            /wiki/Manne_Siegbahn   
James Franck                                                /wiki/James_Franck   
Gustav Hertz                                         /wiki/Gustav_Ludwig_Hertz   
...                                                                        ...   
John C. Mather                                            /wiki/John_C._Mather   
George F. Smoot                                             /wiki/George_Smoot   
Albert Fert                                                  /wiki/Albert_Fert   
Peter Grünberg                                       /wiki/Peter_Gr%C3%BCnberg   
Makoto Kobayashi                            /wiki/Makoto_Kobayashi_(physicist)   
Toshihide Maskawa                                      /wiki/Toshihide_Maskawa   
Yoichiro Nambu                                            /wiki/Yoichiro_Nambu   
Charles K. Kao                                            /wiki/Charles_K._Kao   
Willard S. Boyle                                           /wiki/Willard_Boyle   
George E. Smith                                          /wiki/George_E._Smith   
Andre Geim                                                    /wiki/Andre_Geim   
Konstantin Novoselov                                /wiki/Konstantin_Novoselov   
Saul Perlmutter                                          /wiki/Saul_Perlmutter   
Brian P. Schmidt                                           /wiki/Brian_Schmidt   
Adam G. Riess                                                 /wiki/Adam_Riess   
Serge Haroche                                              /wiki/Serge_Haroche   
David J. Wineland                                      /

Next, we gather the information that we really want: The number of words in each Wikipedia entry that are in the physics corpus, and store them in a new column in both dataframes (the column is named "length").

- ### Compute the length of each list

In [37]:
nobel_df["length"] =  nobel_df["physics_list_clean"].apply(len)
physics_df["length"] = physics_df["physics_list_clean"].apply(len)

Lastly, we are eventually going to keep all this information in a single dataframe. However, we still want to identify which entries are physicists, and which are physics domains. So we create a new column, named "group", in both dataframes. Entries in nobel_df receive a value of 1 for "group", while those in physics_df are set to 0.

- ### Create "group" column: 1 for nobel_df, 0 for physics_df

In [38]:
nobel_df["group"] =  1

In [39]:
physics_df["group"] =  0

Now that we have essentialy all the information we need, we simply need to put it into the right format. Recall that a social network consists of nodes and links connecting those nodes. Specifically, we are going to create a list containing the nodes (nodes_list), and a df containing the links (links_list). For now, we will focus on creating the nodes_list.

## Create nodes_list

nodes_list: A list of dictionaries, each containing a physics domain/physicist, the length of their Wikipedia entry (specifically, how many words in their entry were also in physics_corpus), and their group (0 if it is a physics domain, 1 if it is a physicist).

First, let's get the relevant info for all the physicists and physics domains into one dataframe.

- ### Concatenate phsycics_df and nobel_df to form nodes_df

In [40]:
nodes_df =  pd.concat([physics_df, nobel_df])[['length','group']]

nodes_df.index.name = "id"
nodes_df

length  group
id                                                   
accelerator physics                        866      0
acoustics                                 2072      0
agrophysics                                398      0
antimatter                                3532      0
applied physics                            289      0
astrometry                                1151      0
astronomy                                 5207      0
astrophysics                              1859      0
atom                                      6633      0
atomic and molecular astrophysics          471      0
atomic physics                             964      0
atomic, molecular, and optical physics    1870      0
bcs theory                                1362      0
big bang                                  6932      0
biophysics                                1074      0
black hole                                8355      0
bloch wave                                 839      0
bose-einstein condensate                  3494      0
brane                                      105      0
casimir effect                            3105      0
chemical physics                           496      0
communication physics                      259      0
condensed matter physics                  3486      0
cosmic background radiation                517      0
cosmic inflation                          6381      0
cosmic string                             1610      0
cosmology                                 4294      0
cosmos                                     769      0
dark energy                               4323      0
dark matter                               6844      0
...                                        ...    ...
John C. Mather                            1180      1
George F. Smoot                           1679      1
Albert Fert                                709      1
Peter Grünberg                             879      1
Makoto Kobayashi                           675      1
Toshihide Maskawa                          609      1
Yoichiro Nambu                            1636      1
Charles K. Kao                            4088      1
Willard S. Boyle                           709      1
George E. Smith                            544      1
Andre Geim                                2205      1
Konstantin Novoselov                      1446      1
Saul Perlmutter                           1388      1
Brian P. Schmidt                          2385      1
Adam G. Riess                             1221      1
Serge Haroche                              865      1
David J. Wineland                         1481      1
François Englert                          1130      1
Peter Higgs                               2496      1
Isamu Akasaki                             1277      1
Hiroshi Amano                              594      1
Shuji Nakamura                             947      1
Takaaki Kajita                            1006      1
Arthur B. McDonald                        1166      1
David J. Thouless                          934      1
F. Duncan M. Haldane                       813      1
John M. Kosterlitz                         621      1
Rainer Weiss                              1859      1
Kip Thorne                                2851      1
Barry Barish                              1466      1

[339 rows x 2 columns]

Since we are using the D3.js library, we must properly format as a list of dictionaries. Each dictionary will contain the group, length, and name of the physicist/physics domain (as 'id').

- ### Convert nodes_df to list of dictionaries

In [41]:
nodes_list = list(nodes_df.reset_index().transpose().to_dict().values())
nodes_list

[{'group': 0, 'id': 'accelerator physics', 'length': 866},
 {'group': 0, 'id': 'acoustics', 'length': 2072},
 {'group': 0, 'id': 'agrophysics', 'length': 398},
 {'group': 0, 'id': 'antimatter', 'length': 3532},
 {'group': 0, 'id': 'applied physics', 'length': 289},
 {'group': 0, 'id': 'astrometry', 'length': 1151},
 {'group': 0, 'id': 'astronomy', 'length': 5207},
 {'group': 0, 'id': 'astrophysics', 'length': 1859},
 {'group': 0, 'id': 'atom', 'length': 6633},
 {'group': 0, 'id': 'atomic and molecular astrophysics', 'length': 471},
 {'group': 0, 'id': 'atomic physics', 'length': 964},
 {'group': 0, 'id': 'atomic, molecular, and optical physics', 'length': 1870},
 {'group': 0, 'id': 'bcs theory', 'length': 1362},
 {'group': 0, 'id': 'big bang', 'length': 6932},
 {'group': 0, 'id': 'biophysics', 'length': 1074},
 {'group': 0, 'id': 'black hole', 'length': 8355},
 {'group': 0, 'id': 'bloch wave', 'length': 839},
 {'group': 0, 'id': 'bose-einstein condensate', 'length': 3494},
 {'group': 0

Creating nodes_list was easy. Creating links_list will be more involved.

## Create links_list

Ultimately we want to create a list of libraries, each containing two physicists/physics domains (one listed as "target", the other as "source"), and their "value", which measures the strength of their connection (we will discuss later how the strength of their connection is computed).

Here is the basic outline of what we are going to do:

- Create a df (words_vector) containing the how many times each word in physics_corpus occurs in each physicist/physics domain entry.
- Create another df (similarity_df) containing the cosine-similarity index between each of two entries. This is the value we will use to measure how "connected" two physicists/physics domains are. (The cosine-similarity index will be explained later)
- Melt similarity_df into a df containing three columns ("source", "target", "value". The resulting df is named melted_df.
- Convert similarity_df into a list of dictionaries. This list will be named links_list.

Once we have that, we can use nodes_list and links_list to visualize our social network.

Let's start with creating words_vector. We will create an empty dataframe, populate it with the count of each word per Wikipedia entry, and then fill in any missing values with a 0.

- ### Create a(n empty) data frame with the nodes_df index as columns and physics_corpus as index

In [42]:
words_vector =  pd.DataFrame( columns = nodes_df.index.values, index = physics_corpus)
words_vector

accelerator physics acoustics agrophysics antimatter  \
paved                                NaN       NaN         NaN        NaN   
proposal                             NaN       NaN         NaN        NaN   
released                             NaN       NaN         NaN        NaN   
wwwencyclopediacom                   NaN       NaN         NaN        NaN   
pbar                                 NaN       NaN         NaN        NaN   
exact                                NaN       NaN         NaN        NaN   
mystic                               NaN       NaN         NaN        NaN   
thickness                            NaN       NaN         NaN        NaN   
element                              NaN       NaN         NaN        NaN   
aas                                  NaN       NaN         NaN        NaN   
poisonous                            NaN       NaN         NaN        NaN   
ralph                                NaN       NaN         NaN        NaN   
oleary                               NaN       NaN         NaN        NaN   
distorted                            NaN       NaN         NaN        NaN   
aqueous                              NaN       NaN         NaN        NaN   
howard                               NaN       NaN         NaN        NaN   
society                              NaN       NaN         NaN        NaN   
reject                               NaN       NaN         NaN        NaN   
le                                   NaN       NaN         NaN        NaN   
ramsey                               NaN       NaN         NaN        NaN   
students                             NaN       NaN         NaN        NaN   
liberated                            NaN       NaN         NaN        NaN   
jaguar                               NaN       NaN         NaN        NaN   
mother                               NaN       NaN         NaN        NaN   
wrinkles                             NaN       NaN         NaN        NaN   
kids                                 NaN       NaN         NaN        NaN   
longest                              NaN       NaN         NaN        NaN   
europhysics                          NaN       NaN         NaN        NaN   
electrochemical                      NaN       NaN         NaN        NaN   
abelian                              NaN       NaN         NaN        NaN   
...                                  ...       ...         ...        ...   
lamb                                 NaN       NaN         NaN        NaN   
disagree                             NaN       NaN         NaN        NaN   
xn                                   NaN       NaN         NaN        NaN   
homestake                            NaN       NaN         NaN        NaN   
traverses                            NaN       NaN         NaN        NaN   
bergman                              NaN       NaN         NaN        NaN   
powerpoint                           NaN       NaN         NaN        NaN   
broglie                              NaN       NaN         NaN        NaN   
gammarays                            NaN       NaN         NaN        NaN   
arxivquantph                         NaN       NaN         NaN        NaN   
flowing                              NaN       NaN         NaN        NaN   
sensible                             NaN       NaN         NaN        NaN   
scholars                             NaN       NaN         NaN        NaN   
compendium                           NaN       NaN         NaN        NaN   
quantentheoretischen                 NaN       NaN         NaN        NaN   
efforts                              NaN       NaN         NaN        NaN   
pulsating                            NaN       NaN         NaN        NaN   
kluwer                               NaN       NaN         NaN        NaN   
gl                                   NaN       NaN         NaN        NaN   
thermionic                           NaN       NaN         NaN        NaN   
unifying         

Now we need to make a function which takes a list, and returns the count for each element in the list. The format will be a Series with the elements of the list as index, and contains the count of each one.

We will apply this function to each element in "physics_list_clean" in both nobel_df  and physics_df. Recall that physics_list_clean columns contains all words from each physicist's/physics domain's Wikipedia entry that are also physics_corpus. Thus, applying this function to an element of "physics_list_clean" will return a Series, with index physics_corpus and counts for each. Thus, it will work perfectly to fill each column of words_vector.

- ### Write/apply function that returns word count of given list

In [43]:
def count_words(list_to_count):
    return (pd.Series(list_to_count)).value_counts()

words_vector.loc[:,nobel_df.index] = nobel_df["physics_list_clean"].apply(count_words).transpose()
words_vector.loc[:,physics_df.index] = physics_df["physics_list_clean"].apply(count_words).transpose()

Now we replace any NaN values with 0.

- ### Fill the missing values

In [44]:
words_vector = words_vector.fillna(value=0)
words_vector
#nobel_df["physics_list_clean"].apply(count_words)

accelerator physics  acoustics  agrophysics  antimatter  \
paved                                 0.0        0.0          0.0         0.0   
proposal                              0.0        0.0          0.0         1.0   
released                              0.0        0.0          0.0         3.0   
wwwencyclopediacom                    0.0        0.0          0.0         0.0   
pbar                                  0.0        0.0          0.0         0.0   
exact                                 1.0        0.0          1.0         1.0   
mystic                                0.0        0.0          0.0         0.0   
thickness                             0.0        0.0          0.0         0.0   
element                               0.0        1.0          0.0         0.0   
aas                                   0.0        0.0          0.0         0.0   
poisonous                             0.0        0.0          0.0         0.0   
ralph                                 0.0        0.0          0.0         0.0   
oleary                                0.0        0.0          0.0         0.0   
distorted                             0.0        0.0          0.0         0.0   
aqueous                               0.0        0.0          0.0         0.0   
howard                                0.0        0.0          0.0         0.0   
society                               0.0       16.0          2.0         5.0   
reject                                0.0        0.0          0.0         0.0   
le                                    0.0        0.0          0.0         0.0   
ramsey                                0.0        0.0          0.0         0.0   
students                              0.0        0.0          0.0         0.0   
liberated                             0.0        0.0          0.0         1.0   
jaguar                                0.0        0.0          0.0         0.0   
mother                                0.0        0.0          0.0         0.0   
wrinkles                              0.0        0.0          0.0         0.0   
kids                                  0.0        0.0          0.0         0.0   
longest                               0.0        0.0          0.0         0.0   
europhysics                           0.0        0.0          0.0         0.0   
electrochemical                       0.0        0.0          0.0         0.0   
abelian                               0.0        0.0          0.0         0.0   
...                                   ...        ...          ...         ...   
lamb                                  0.0        1.0          0.0         0.0   
disagree                              0.0        0.0          0.0         0.0   
xn                                    0.0        0.0          0.0         0.0   
homestake                             0.0        0.0          0.0         0.0   
traverses                             0.0        0.0          0.0         0.0   
bergman                               0.0        0.0          0.0         0.0   
powerpoint                            0.0        0.0          0.0         0.0   
broglie                               0.0        0.0          0.0         2.0   
gammarays                             0.0        0.0          0.0         0.0   
arxivquantph                          0.0        0.0          0.0         0.0   
flowing                               0.0        0.0          0.0         0.0   
sensible                              0.0        0.0          0.0         0.0   
scholars                              0.0        0.0          0.0         0.0   
compendium                            0.0        0.0          0.0         0.0   
quantentheoretischen                  0.0        0.0          0.0         0.0   
efforts                               0.0        0.0          0.0         0.0   
pulsating                             0.0        0.0          0.0         0.0   
kluwer                                0.0        0.0 

Next, we want to create similarity_df. Before we can, we must define the cosine similarity index, and compute it for each pair of physicists/physics domains.

According to Wikipedia, the "Cosine similarity is a measure of similarity between two non-zero vectors of an inner product space that measures the cosine of the angle between them." In our situation, we will use the words in physics_corpus as an orthogonal basis for the vectors in our space. These vectors represent each physicist/physics domain Wikipedia entry. Specifically, each vector will be 12935 elements long, each element being the number of times the corresponding physics_corpus word appears in the vector (AKA Wikipedia entry). A cosine similarity index of 1 means that the two entries are in the exact same direction. A value of -1 means that they are in exact opposite direction.

Thus for any two Wikipedia entries, we can compute the cosine similarity index as

$$similarity = cos\left(\theta\right) = \frac{A\cdot B}{\left\|A\right\| \left\|B\right\|},$$ where $A, B$ are their vector representations. A similarity value of 1 means that the two entries are completely identical. A value of -1 means that they are completely opposite.



- ### Create similarity_df

In [63]:
def euc_norm(list):
    return np.sqrt( sum([x**2 for x in list]))
    
# Use "norm"s for denominator
norm = pd.DataFrame(words_vector.apply(euc_norm))
    
numer = words_vector.transpose().dot(words_vector)
denom = norm.dot( norm.transpose() )
similarity_df = numer/denom

In [64]:
similarity_df

accelerator physics  acoustics  \
accelerator physics                                1.000000   0.181027   
acoustics                                          0.181027   1.000000   
agrophysics                                        0.324358   0.232904   
antimatter                                         0.243108   0.145087   
applied physics                                    0.476278   0.204141   
astrometry                                         0.133262   0.137180   
astronomy                                          0.169541   0.179164   
astrophysics                                       0.296805   0.224255   
atom                                               0.250150   0.163213   
atomic and molecular astrophysics                  0.106064   0.064637   
atomic physics                                     0.370747   0.167784   
atomic, molecular, and optical physics             0.380099   0.206760   
bcs theory                                         0.130958   0.098204   
big bang                                           0.148255   0.112224   
biophysics                                         0.209139   0.159056   
black hole                                         0.119595   0.088073   
bloch wave                                         0.105065   0.138314   
bose-einstein condensate                           0.207098   0.103820   
brane                                              0.042053   0.051464   
casimir effect                                     0.231781   0.133763   
chemical physics                                   0.290227   0.124739   
communication physics                              0.202803   0.139637   
condensed matter physics                           0.388740   0.206765   
cosmic background radiation                        0.081196   0.054156   
cosmic inflation                                   0.167451   0.111732   
cosmic string                                      0.072375   0.100303   
cosmology                                          0.211602   0.150225   
cosmos                                             0.099352   0.110556   
dark energy                                        0.114450   0.094833   
dark matter                                        0.140769   0.085197   
...                                                     ...        ...   
John C. Mather                                     0.111033   0.086052   
George F. Smoot                                    0.147118   0.105797   
Albert Fert                                        0.141381   0.079228   
Peter Grünberg                                     0.118677   0.070847   
Makoto Kobayashi                                   0.239615   0.110185   
Toshihide Maskawa                                  0.191127   0.095180   
Yoichiro Nambu                                     0.152042   0.114350   
Charles K. Kao                                     0.136672   0.156477   
Willard S. Boyle                                   0.113726   0.079910   
George E. Smith                                    0.097994   0.075364   
Andre Geim                                         0.177628   0.132179   
Konstantin Novoselov                               0.196271   0.151797   
Saul Perlmutter                                    0.112366   0.068587   
Brian P. Schmidt                                   0.113383   0.104419   
Adam G. Riess                                      0.106271   0.082070   
Serge Haroche                                      0.144678   0.095556   
David J. Wineland                                  0.127581   0.109672   
François Englert                                   0.238962   0.114421   
Peter Higgs                                        0.198304   0.109187   
Isamu Akasaki                                      0.154898   0.140021   
Hiroshi Amano                                      0.172038   0.115521   
Shuji Nakamura                                     0.120158   0.092485   
Takaaki Kajita                              

We have all the information we need! We have the cosine similarity for every pair of physicists/physics domains, but we're going to clean it up some more. Here's what we need to do.

- Melt the df so that the data is easier to work with.
- Get rid of duplicates (for example, there is a row with acoustics for "source" and accelerator physics for "target". There is also a row with accelerator physics for "source" and aocustics for "target". Because our social network is not directed, we treat these as the same. Thus we are going to only keep one of these entries. Similarly, we will remove all other "duplicates"
- We want to highlight only the most important connections. So for every physicist/physics domain, we only keep the 8 physicists/domains with the highest cosine similarity index to it.


- ### Reset the index and melt the dataframe

In [48]:
melted_df = pd.melt(similarity_df.reset_index(), id_vars = ['index'])
melted_df.columns = ["source", "target", "value"]
melted_df

source               target     value
0                          accelerator physics  accelerator physics  1.000000
1                                    acoustics  accelerator physics  0.181027
2                                  agrophysics  accelerator physics  0.324358
3                                   antimatter  accelerator physics  0.243108
4                              applied physics  accelerator physics  0.476278
5                                   astrometry  accelerator physics  0.133262
6                                    astronomy  accelerator physics  0.169541
7                                 astrophysics  accelerator physics  0.296805
8                                         atom  accelerator physics  0.250150
9            atomic and molecular astrophysics  accelerator physics  0.106064
10                              atomic physics  accelerator physics  0.370747
11      atomic, molecular, and optical physics  accelerator physics  0.380099
12                                  bcs theory  accelerator physics  0.130958
13                                    big bang  accelerator physics  0.148255
14                                  biophysics  accelerator physics  0.209139
15                                  black hole  accelerator physics  0.119595
16                                  bloch wave  accelerator physics  0.105065
17                    bose-einstein condensate  accelerator physics  0.207098
18                                       brane  accelerator physics  0.042053
19                              casimir effect  accelerator physics  0.231781
20                            chemical physics  accelerator physics  0.290227
21                       communication physics  accelerator physics  0.202803
22                    condensed matter physics  accelerator physics  0.388740
23                 cosmic background radiation  accelerator physics  0.081196
24                            cosmic inflation  accelerator physics  0.167451
25                               cosmic string  accelerator physics  0.072375
26                                   cosmology  accelerator physics  0.211602
27                                      cosmos  accelerator physics  0.099352
28                                 dark energy  accelerator physics  0.114450
29                                 dark matter  accelerator physics  0.140769
...                                        ...                  ...       ...
114891                          John C. Mather         Barry Barish  0.372650
114892                         George F. Smoot         Barry Barish  0.352798
114893                             Albert Fert         Barry Barish  0.393766
114894                          Peter Grünberg         Barry Barish  0.301966
114895                        Makoto Kobayashi         Barry Barish  0.400255
114896                       Toshihide Maskawa         Barry Barish  0.411192
114897                          Yoichiro Nambu         Barry Barish  0.436922
114898                          Charles K. Kao         Barry Barish  0.382160
114899                        Willard S. Boyle         Barry Barish  0.380756
114900                         George E. Smith         Barry Barish  0.411351
114901                              Andre Geim         Barry Barish  0.538539
114902                    Konstantin Novoselov         Barry Barish  0.468755
114903                         Saul Perlmutter         Barry Barish  0.379704
114904                        Brian P. Schmidt         Barry Barish  0.376633
114905                           Adam G. Riess         Barry Barish  0.394487
114906                           Serge Haroche         Barry Barish  0.381999
114907                       David J. Wineland         Barry Barish  0.446275
114908                        François Englert         Barry Barish  0.438085
114909                             Peter Higgs         Barry Barish  0.327415
114910                           Isamu Akasaki         Barry B

- ### Shuffle data set rowwise.

In [49]:
melted_df = melted_df.sample(frac=1.).reset_index(drop=True)
melted_df

source                          target  \
0                              antimatter         Rudolf Ludwig Mössbauer   
1                     Joseph John Thomson         William Daniel Phillips   
2        nanoscale and mesoscopic physics          physics of computation   
3                              Kip Thorne             gravitation physics   
4                             Peter Higgs        Albert Abraham Michelson   
5                              biophysics             Luis Walter Alvarez   
6                              Steven Chu                      Steven Chu   
7                         nuclear physics                   Tsung-Dao Lee   
8                         applied physics            Horst Ludwig Störmer   
9                         Georges Charpak              Arthur B. McDonald   
10                     Willis Eugene Lamb           high pressure physics   
11                   neutrino oscillation            nuclear astrophysics   
12                   theory of everything                      Max Planck   
13                         Alfred Kastler                          string   
14                          cosmic string               Robert Hofstadter   
15               Joseph Hooton Taylor Jr.              Simon van der Meer   
16        Patrick Maynard Stuart Blackett                      superfluid   
17                                 galaxy           Lev Davidovich Landau   
18                          Wolfgang Paul         quantum electrodynamics   
19                          Frits Zernike         William Daniel Phillips   
20                          Serge Haroche                    Enrico Fermi   
21                    Charles Hard Townes            quantum field theory   
22                     John M. Kosterlitz         low-temperature physics   
23                            Abdus Salam                          optics   
24                  Robert Woodrow Wilson         William Daniel Phillips   
25                  William Alfred Fowler                         nuclear   
26                        Herbert Kroemer       William Bradford Shockley   
27                 Philip Warren Anderson                  Polykarp Kusch   
28                         Charles K. Kao          Edward Victor Appleton   
29                              Leo Esaki             Alexander Prokhorov   
...                                   ...                             ...   
114891                      Hiroshi Amano                       magnetism   
114892                      photovoltaics             William Henry Bragg   
114893                      semiconductor                   Serge Haroche   
114894             Maria Skłodowska-Curie             Victor Francis Hess   
114895                     Roy J. Glauber                 Albert Einstein   
114896                  Wolfgang Ketterle             accelerator physics   
114897                James Watson Cronin                             gas   
114898                          photonics                  James Chadwick   
114899             Claude Cohen-Tannoudji                            spin   
114900                      Isamu Akasaki              Bertram Brockhouse   
114901                 John M. Kosterlitz                  Peter Grünberg   
114902                             liquid            Norman Foster Ramsey   
114903    Patrick Maynard Stuart Blackett                  John C. Mather   
114904                          astronomy     Charles Thomson Rees Wilson   
114905             Philip Warren Anderson           electrical conduction   
114906                 Robert B. Laughlin                      Louis Néel   
114907                  molecular physics                   photovoltaics   
114908                  Guglielmo Marconi                Owen Chamberlain   
114909                engineering physics             Luis Walter Alvarez   
114910                  self-organization                  James Chadwick   
114911                Carl Davi

Getting rid of duplicates is a bit tricky. What we do is merge melted_df with itself, merging on ['source', 'target'] with ['target', 'source']. We keep the indices of both "copies" of melted_df. For each pair of duplicates (i.e. source_x, target_x =target_x, source_y),  we store in the list "index_to_drop" the value of either index_x or index_y, whichever is greater (using .unique() so that we don't have duplicates in this list). Finally, we drop all indices in "index_to_drop" from melted_df to get melted_df_sub, which has no duplicates.

- ### Merge melted_df with itself

In [50]:
merged_df = pd.merge(melted_df.reset_index(), melted_df.reset_index(), left_on =['source','target'], right_on=['target','source'])

merged_df

index_x                           source_x  \
0             0                         antimatter   
1             1                Joseph John Thomson   
2             2   nanoscale and mesoscopic physics   
3             3                         Kip Thorne   
4             4                        Peter Higgs   
5             5                         biophysics   
6             6                         Steven Chu   
7             7                    nuclear physics   
8             8                    applied physics   
9             9                    Georges Charpak   
10           10                 Willis Eugene Lamb   
11           11               neutrino oscillation   
12           12               theory of everything   
13           13                     Alfred Kastler   
14           14                      cosmic string   
15           15           Joseph Hooton Taylor Jr.   
16           16    Patrick Maynard Stuart Blackett   
17           17                             galaxy   
18           18                      Wolfgang Paul   
19           19                      Frits Zernike   
20           20                      Serge Haroche   
21           21                Charles Hard Townes   
22           22                 John M. Kosterlitz   
23           23                        Abdus Salam   
24           24              Robert Woodrow Wilson   
25           25              William Alfred Fowler   
26           26                    Herbert Kroemer   
27           27             Philip Warren Anderson   
28           28                     Charles K. Kao   
29           29                          Leo Esaki   
...         ...                                ...   
114891   114891                      Hiroshi Amano   
114892   114892                      photovoltaics   
114893   114893                      semiconductor   
114894   114894             Maria Skłodowska-Curie   
114895   114895                     Roy J. Glauber   
114896   114896                  Wolfgang Ketterle   
114897   114897                James Watson Cronin   
114898   114898                          photonics   
114899   114899             Claude Cohen-Tannoudji   
114900   114900                      Isamu Akasaki   
114901   114901                 John M. Kosterlitz   
114902   114902                             liquid   
114903   114903    Patrick Maynard Stuart Blackett   
114904   114904                          astronomy   
114905   114905             Philip Warren Anderson   
114906   114906                 Robert B. Laughlin   
114907   114907                  molecular physics   
114908   114908                  Guglielmo Marconi   
114909   114909                engineering physics   
114910   114910                  self-organization   
114911   114911                Carl David Anderson   
114912   114912                      spectral line   
114913   114913                Victor Francis Hess   
114914   114914       Pavel Alekseyevich Cherenkov   
114915   114915                     standard model   
114916   114916              electrical conduction   
114917   114917                        Gerd Binnig   
114918   114918  atomic and molecular astrophysics   
114919   114919                      laser physics   
114920   114920                  Toshihide Maskawa   

                              target_x   value_x  index_y  \
0              Rudolf Ludwig Mössbauer  0.134750    34893   
1              William Daniel Phillips  0.233942     2597   
2               physics of computation  0.079613    59382   
3                  gravitation physics  0.344277    51995   
4             Albert Abraham Michelson  0.206263    63381   
5                  Luis Walter Alvarez  0.085663    10652   
6                           Steven Chu  1.000000        6   
7                        Tsung-Dao Lee  0.206987    88144   
8                 Horst Ludwig Störmer  0.220484    49847   
9                   Arthur B. McDonald  0.373899 

- ### Find the indices to drop

In [51]:
index_to_drop = merged_df[["index_x", "index_y"]].apply(max, axis=1).unique()

- ### Use the index_to_drop to subset the melted_df dataframe

In [52]:
melted_df_sub =  melted_df.drop(index_to_drop)
melted_df_sub

source                     target     value
0                             antimatter    Rudolf Ludwig Mössbauer  0.134750
1                    Joseph John Thomson    William Daniel Phillips  0.233942
2       nanoscale and mesoscopic physics     physics of computation  0.079613
3                             Kip Thorne        gravitation physics  0.344277
4                            Peter Higgs   Albert Abraham Michelson  0.206263
5                             biophysics        Luis Walter Alvarez  0.085663
7                        nuclear physics              Tsung-Dao Lee  0.206987
8                        applied physics       Horst Ludwig Störmer  0.220484
9                        Georges Charpak         Arthur B. McDonald  0.373899
10                    Willis Eugene Lamb      high pressure physics  0.069202
11                  neutrino oscillation       nuclear astrophysics  0.294435
12                  theory of everything                 Max Planck  0.214999
13                        Alfred Kastler                     string  0.055601
14                         cosmic string          Robert Hofstadter  0.045146
15              Joseph Hooton Taylor Jr.         Simon van der Meer  0.303808
16       Patrick Maynard Stuart Blackett                 superfluid  0.089802
17                                galaxy      Lev Davidovich Landau  0.079503
18                         Wolfgang Paul    quantum electrodynamics  0.107877
19                         Frits Zernike    William Daniel Phillips  0.442416
20                         Serge Haroche               Enrico Fermi  0.207518
21                   Charles Hard Townes       quantum field theory  0.097739
22                    John M. Kosterlitz    low-temperature physics  0.035858
23                           Abdus Salam                     optics  0.124684
24                 Robert Woodrow Wilson    William Daniel Phillips  0.326109
25                 William Alfred Fowler                    nuclear  0.229523
26                       Herbert Kroemer  William Bradford Shockley  0.305241
27                Philip Warren Anderson             Polykarp Kusch  0.264253
28                        Charles K. Kao     Edward Victor Appleton  0.201653
29                             Leo Esaki        Alexander Prokhorov  0.357860
30                    Klaus von Klitzing       stellar astrophysics  0.096349
...                                  ...                        ...       ...
112482                      econophysics    William Daniel Phillips  0.119236
112530                            phonon               Carlo Rubbia  0.143899
112540         Pyotr Leonidovich Kapitsa            Steven Weinberg  0.255911
112541                              atom             Roy J. Glauber  0.222312
112563                    Chen-Ning Yang                 Andre Geim  0.395182
112569                        astrometry          materials physics  0.158080
112588               engineering physics              semiconductor  0.147767
112642                         Aage Bohr    quantum electrodynamics  0.118144
112678                        Otto Stern     John Douglas Cockcroft  0.244912
112765                        Max Planck                 black hole  0.121911
112786                  Owen Chamberlain                     planet  0.053535
112922          Hannes Olof Gösta Alfvén             Chen-Ning Yang  0.237159
112930                    Roy J. Glauber     quantum chromodynamics  0.210943
112963                      solar system           Daniel Chee Tsui  0.043379
112974              neutrino oscillation          Martin Lewis Perl  0.169766
113093                        Paul Dirac                 bcs theory  0.197078
113110                            liquid           Murray Gell-Mann  0.045978
113158                         supernova                   molecule  0.126103
113268                physical chemistry            Ernest Lawrence  0.093681
113316       cosmic background radiation        accelera

And now, for every physicist/domain, we keep only the 8 strongest connections. We store this in largest_df, which will be a 2-level multiindex. We capture the values of index 1 and store them in index_to_keep. These store the indices we want to keep in our links_df.

- ### Group melted_df_sub by "source" using the groupby method and select the 8 targets with highest "value."

In [66]:
largest_df = melted_df_sub.groupby('source')['value'].nlargest(8)

- ### Get the level 1 of the multiindex

In [67]:
index_to_keep =  largest_df.index.get_level_values(1)

links_df = melted_df_sub.loc[index_to_keep]
links_df

source                                 target  \
1514                   Aage Bohr                             Niels Bohr   
9442                   Aage Bohr                      Ben Roy Mottelson   
22853                  Aage Bohr                    Leo James Rainwater   
87695                  Aage Bohr                   Nicolaas Bloembergen   
32066                  Aage Bohr                           James Franck   
4246                   Aage Bohr                     Jerome I. Friedman   
2537                   Aage Bohr                             Andre Geim   
79156                  Aage Bohr                      David J. Thouless   
58248                Abdus Salam                             Andre Geim   
82062                Abdus Salam                   Nicolaas Bloembergen   
31096                Abdus Salam                  Anthony James Leggett   
16476                Abdus Salam                       Makoto Kobayashi   
54470                Abdus Salam                     Arthur B. McDonald   
32080                Abdus Salam                     Bertram Brockhouse   
50668                Abdus Salam                        applied physics   
18331                Abdus Salam                         David J. Gross   
68030              Adam G. Riess                       Brian P. Schmidt   
20712              Adam G. Riess                         Takaaki Kajita   
46781              Adam G. Riess                           Rainer Weiss   
1967               Adam G. Riess                             Andre Geim   
18689              Adam G. Riess                      David J. Thouless   
80502              Adam G. Riess                      Masatoshi Koshiba   
83223              Adam G. Riess                   Horst Ludwig Störmer   
14104              Adam G. Riess                     Bertram Brockhouse   
10658   Albert Abraham Michelson                   Norman Foster Ramsey   
2074    Albert Abraham Michelson                 John Robert Schrieffer   
7494    Albert Abraham Michelson                   Nicolaas Bloembergen   
7029    Albert Abraham Michelson                Percy Williams Bridgman   
1957    Albert Abraham Michelson                     Jerome I. Friedman   
29606   Albert Abraham Michelson                Arthur Leonard Schawlow   
...                          ...                                    ...   
90078                   universe                            dark matter   
11830                   universe            cosmic background radiation   
29298                   universe                   stellar astrophysics   
16584                   universe                  Robert Woodrow Wilson   
77118                   universe                             antimatter   
68931                   universe                     general relativity   
9325               vacuum energy                         casimir effect   
3581               vacuum energy                    quantum electronics   
61255              vacuum energy                        quantum gravity   
37774              vacuum energy                      quantum chemistry   
279                vacuum energy                                 photon   
77097              vacuum energy                       particle physics   
49577              vacuum energy                         atomic physics   
12344              vacuum energy                   theory of everything   
56583           vehicle dynamics                         fluid dynamics   
103579          vehicle dynamics                      self-organization   
3406            vehicle dynamics  newton's law of universal gravitation   
66844           vehicle dynamics                        electromagnetic   
4692            vehicle dynamics                        polymer physics   
35220           vehicle dynamics                                    gas   
5897            vehicle dynamics                       seesaw mechanism   
89240           vehicle dynamics                   

Finally, we conver this to a list.

- ### Create the links_list

In [68]:
links_list =  list(links_df.transpose().to_dict().values())
links_list

[{'source': 'Aage Bohr', 'target': 'Niels Bohr', 'value': 0.810448278052767},
 {'source': 'Aage Bohr',
  'target': 'Ben Roy Mottelson',
  'value': 0.6405523849120306},
 {'source': 'Aage Bohr',
  'target': 'Leo James Rainwater',
  'value': 0.59363610094523},
 {'source': 'Aage Bohr',
  'target': 'Nicolaas Bloembergen',
  'value': 0.4535685670272227},
 {'source': 'Aage Bohr',
  'target': 'James Franck',
  'value': 0.45121833589128957},
 {'source': 'Aage Bohr',
  'target': 'Jerome I. Friedman',
  'value': 0.4292465753445105},
 {'source': 'Aage Bohr', 'target': 'Andre Geim', 'value': 0.42499050449712006},
 {'source': 'Aage Bohr',
  'target': 'David J. Thouless',
  'value': 0.4031941944705469},
 {'source': 'Abdus Salam',
  'target': 'Andre Geim',
  'value': 0.3080747544129564},
 {'source': 'Abdus Salam',
  'target': 'Nicolaas Bloembergen',
  'value': 0.2988046873651129},
 {'source': 'Abdus Salam',
  'target': 'Anthony James Leggett',
  'value': 0.2873736808741804},
 {'source': 'Abdus Salam',

Similar to how we did at the beginning of this project for the small network, we export this as a json file. As before, we dump this file in a folder named "data", which is itself inside a folder containing "index.html". Open up index.html, and you will be able to interact with our social network. 

When you open up this file, you will see a social network. Each circle represents either a physicist ("red") or domain ("blue"). Additionally, clicking on one of them will highlight the 8 closest connections. The size of each circle is determined by how many words were in the corresponding Wikipedia entry.

In [69]:
network_dict = {"nodes": nodes_list,
                "links": links_list}

with open(".\data\physicists.json","w") as f:
    json.dump(network_dict, f, indent=4)